src/movies.js

```diff
@@ -1,30 +1,110 @@
-function getAllDirectors(moviesArray) {}
+function getAllDirectors(moviesArray) {
+  let allDirectors = moviesArray.map((movie) => {
+    return movie.director;
+  });
+  return allDirectors;
+}
 
 // Iteration 2: Steven Spielberg. The best? - How many drama movies did STEVEN SPIELBERG direct?
-function howManyMovies(moviesArray) {}
+function howManyMovies(moviesArray) {
+  return moviesArray.filter((movie) => {
+    return (
+      movie.director === 'Steven Spielberg' && movie.genre.includes('Drama')
+    );
+  }).length;
+}
 
 // Iteration 3: All scores average - Get the average of all scores with 2 decimals
-function scoresAverage(moviesArray) {}
+function scoresAverage(moviesArray) {
+  if (moviesArray.length > 0) {
+    let totalScore = moviesArray.reduce((accumulator, currentValue) => {
+      return accumulator + currentValue.score;
+    }, 0);
+    let averageScore = (totalScore / moviesArray.length).toFixed(2);
+    averageScore = Number.parseFloat(averageScore);
+    return averageScore;
+  } else {
+    return 0;
+  }
+}
 
 // Iteration 4: Drama movies - Get the average of Drama Movies
-function dramaMoviesScore(moviesArray) {}
+function dramaMoviesScore(moviesArray) {
+  let dramaMoviesArray = moviesArray.filter((movie) => {
+    return movie.genre.includes('Drama');
+  });
+  if (dramaMoviesArray.length > 0) {
+    let total = dramaMoviesArray.reduce((accumulator, currentValue) => {
+      return accumulator + currentValue.score;
+    }, 0);
+    let averageDramaScore = (total / dramaMoviesArray.length).toFixed(2);
+    averageDramaScore = Number.parseFloat(averageDramaScore);
+    return averageDramaScore;
+  } else {
+    return 0;
+  }
+}
 
 // Iteration 5: Ordering by year - Order by year, ascending (in growing order)
-function orderByYear(moviesArray) {}
+function orderByYear(moviesArray) {
+  const moviesCopy = [...moviesArray];
+  let sorted = moviesCopy.sort((a, b) => {
+    let result = a.year - b.year;
+    if (result !== 0) {
+      return result;
+    }
+    if (a.title < b.title) {
+      return -1;
+    } else if (a.title > b.title) {
+      return 1;
+    } else {
+      return 0;
+    }
+  });
+  return sorted;
+}
 
 // Iteration 6: Alphabetic Order - Order by title and print the first 20 titles
-function orderAlphabetically(moviesArray) {}
+function orderAlphabetically(moviesArray) {
+  let titles = moviesArray.map((movie) => movie.title).sort();
+  let resultTitles = titles.slice(0, 20);
+  return resultTitles;
+}
 
 // BONUS - Iteration 7: Time Format - Turn duration of the movies from hours to minutes
-function turnHoursToMinutes(moviesArray) {}
 
-// BONUS - Iteration 8: Best yearly score average - Best yearly score average
-function bestYearAvg(moviesArray) {}
+// Nic: I did not finish this one, but I did manage to turn hours into minutes, 
+// but it ended up manipulating the original array.. :)  
+function turnHoursToMinutes(moviesArray) {
+    let resultArray = [];
+  moviesArray.forEach((movie) => {
```

Comment:
In this iteration, you firstly want to make deep copy of the array, before you iterate over it.

One approach to make deep copy is this:
```
const newMoviesArry = JSON.parse(JSON.stringify(moviesArray))
```

https://www.freecodecamp.org/news/how-to-clone-an-array-in-javascript-1d3183468f6a/